### PR TITLE
feat(*.ci.jenkins.io) use JDK21 GA (exception of `s390x` sticking to EA)

### PR DESCRIPTION
**Pull request by smerle33**
as per https://github.com/jenkins-infra/helpdesk/issues/3783

paired with @gounthar

### CODE REVIEW

**Comment by lemeurherve** [comment]
I don't have exactly the same here: https://github.com/jenkinsci/docker-agent/pull/521/files#diff-5ccf4f9636b549b76a1ae14b28ab90bbf9a4ba82774e416679d214444cef80ceR33-R35

File: updatecli/scripts/check-jdk.sh
Comment (lemeurherve) [inline updatecli/scripts/check-jdk.sh]: ~~According to https://github.com/adoptium/temurin21-binaries/releases/tag/jdk-21%2B35, the links contain OpenJDK21U-**debugimage**_<...> (ex: `https://github.com/adoptium/temurin21-binaries/releases/download/jdk-21%2B35/OpenJDK21U-debugimage_aarch64_alpine-linux_hotspot_21_35.tar.gz`)~~

~~This will certainly evolve, I would add a TODO about removing it when it will change.~~

File: updatecli/scripts/check-jdk.sh
Comment (lemeurherve) [inline updatecli/scripts/check-jdk.sh]:
```suggestion
      underscoredJDKVersion="${jdk_version//+/_}"
```

File: updatecli/scripts/check-jdk.sh
Comment (lemeurherve) [inline updatecli/scripts/check-jdk.sh]:
```suggestion
      echo "https://github.com/adoptium/temurin21-binaries/releases/download/jdk-${underscoredJDKVersion}/OpenJDK21U-jdk_${platform}_hotspot_${underscoredJDKVersion}"
```

Comment (smerle33) [reply updatecli/scripts/check-jdk.sh]: Why do we need to track those debug image ?

Comment (lemeurherve) [reply updatecli/scripts/check-jdk.sh]: Nevermind, yesterday they were the only ones available, didn't check the link again today.

Comment (lemeurherve) [reply updatecli/scripts/check-jdk.sh]: Edited my suggestion 🙂

Comment (dduportal) [reply updatecli/scripts/check-jdk.sh]: nope, the `debugimage` series is only a subset of all the elements. These are not the binary we want (it was the case for previous JDK btw)

For Linux aarch64 (not Alpine) we want the following:

```
https://github.com/adoptium/temurin21-binaries/releases/download/jdk-21%2B35/OpenJDK21U-jdk_aarch64_linux_hotspot_21_35.tar.gz
```

which means the code in this line looks good

Comment (dduportal) [reply updatecli/scripts/check-jdk.sh]: we don't want to use the debug images. See my comment in https://github.com/jenkins-infra/jenkins-infra/pull/3113#discussion_r1358029572

File: updatecli/scripts/check-jdk.sh
Comment (dduportal) [reply updatecli/scripts/check-jdk.sh]: I remember weird issues when trying to download from the URL with the `_`. Let's keep the encoded version (unless you are able to test your suggestion and prove that it works)

File: updatecli/scripts/check-jdk.sh
Comment (timja) [reply updatecli/scripts/check-jdk.sh]: debug image is a different image and not what we want

Comment (lemeurherve) [reply updatecli/scripts/check-jdk.sh]: Quoting my previous comment:

> Nevermind, yesterday they were the only ones available, didn't check the link again today.

File: updatecli/scripts/check-jdk.sh
Comment (lemeurherve) [inline updatecli/scripts/check-jdk.sh]:
```suggestion
    # No s390x support for JDK21
```
According to https://github.com/adoptium/temurin/issues/8 there won't be any s390x support for JDK21.

File: updatecli/weekly.d/jenkinscontroller-tools-jdk21.yaml
Comment (lemeurherve) [inline updatecli/weekly.d/jenkinscontroller-tools-jdk21.yaml]: NB: these dots in the regexp match any character, not the `.` character.

Not a problem, just a remark.

**Review by lemeurherve — DISMISSED** [review]

File: updatecli/scripts/check-jdk.sh
Comment (gounthar) [reply updatecli/scripts/check-jdk.sh]: Quoting Tim Ellison from Adoptium [in their Slack](https://adoptium.slack.com/archives/C09NLQQAV/p1697121672337069?thread_ts=1697112748.718609&cid=C09NLQQAV):

> As George and Stewart said above, the reason there is no Temurin 21 for ppc64 AIX/s390x Linux/arm32 Linux at the moment is because the OpenJDK 21 GA implementation on those platforms is not passing our release tests. When the fixes are available in OpenJDK source then we can proceed with a release.

So we may see s390x JDK21 builds in the near future (with 21.0.1 for example).
We won't see s390x JDK21+35 builds though.

File: updatecli/weekly.d/jenkinscontroller-tools-jdk21.yaml
Comment (smerle33) [reply updatecli/weekly.d/jenkinscontroller-tools-jdk21.yaml]: we need to match `jdk-21+35` so if we escape the dots, I think we will not catch it, don't you think ?. 
I didn't question the pattern as it is the one used for all our version of jdk but maybe we should.

Comment (lemeurherve) [reply updatecli/weekly.d/jenkinscontroller-tools-jdk21.yaml]: Proposition:
```suggestion
        pattern: "^jdk-21(\\.\\d+)*\\+(\\d+)$"
```
jdk-21 followed by zero to N groups of a dot and one to N numbers, followed by a plus and one to N numbers

Comment (lemeurherve) [reply updatecli/weekly.d/jenkinscontroller-tools-jdk21.yaml]: image

Comment (lemeurherve) [reply updatecli/weekly.d/jenkinscontroller-tools-jdk21.yaml]: I can propose this change in a distinct PR if you prefer.

**Review by gounthar — APPROVED** [review]

File: updatecli/scripts/check-jdk.sh
Comment (dduportal) [inline updatecli/scripts/check-jdk.sh]:
```suggestion
      ## JDK21 URLs have an underscore ('_') instead of a plus ('+') in their archive names
```

Nitpick: Let's keep the same pattern as other JDKs

File: updatecli/scripts/check-jdk.sh
Comment (dduportal) [inline updatecli/scripts/check-jdk.sh:58]: What happens when there is a JDK upgrade not available with the s390x (such as `21+35-ea` => `21+35`)?

- Updatecli still open a PR with the new version in the parameters files (hieradata)
- The build passes as this is only a "hieradata -> puppet -> JCasc YAML file" check
- Then ci.jenkins.io gets updated when the updatecli PR is merged and its "tool" are setup with the new version
- And then, all builds calling the `tool 'jdk21'` pipeline instructions fails as the tool installer responds a HTTP/404 error.


There is no problem on the top-level idea to stop (temporarily) tracking the `s390x` and keep it to a lower version, but it means you must update the tool configuration to keep the old version with a TODO comment as part of the PR here.

**Review by dduportal — CHANGES_REQUESTED** [review]
This is really good job, but we can't merge this "as-it" otherwise we'll break the s390x tool

File: updatecli/scripts/check-jdk.sh
Comment (smerle33) [reply updatecli/scripts/check-jdk.sh:58]: yes but this mean changing the target to point on each of them no ? 
`key: $.profile::jenkinscontroller::jcasc.tools_default_versions.jdk21`

Comment (dduportal) [reply updatecli/scripts/check-jdk.sh:58]: That would work yes. Or you can also keep the updatecli code as it to track and define the default JDK21 version (e.g. `21+35`) and define an exception in the specific hieradata of ci.jenkins.io which defines the s390x agent, as the override mechanism is for the most specific cases like this one.

**Review by dduportal — COMMENTED** [review]
For info, please double check the results as the twin of this PR is failing to update the proper JDK21 download URL when using the new (stable) version in https://github.com/jenkins-infra/packer-images/pull/848#pullrequestreview-1681928106


=> might be packer related only, but worth checking

**Review by dduportal — CHANGES_REQUESTED** [review]
- Updatecli looks to do what is it expected to do:

```
⚠ - key "$.profile::jenkinscontroller::jcasc.tools_default_versions.jdk21" should be updated from "21+35-ea-beta" to "21+35", in file "hieradata/common.yaml"
```

But this PR can't be merged yet:

- The JDK21 URLs must be changed in puppet code (otherwise the PR for `21+35` will break the ci.jenkins.io` tool installation
- The hieradata for ci.jenkins.io does not show specific EA version for s390x

File: hieradata/common.yaml
Comment (dduportal) [inline hieradata/common.yaml:247]:
```suggestion
    jdk21: 21+35-ea-beta
```

**Comment by dduportal** [comment]
Added content to fix the "upgrade" problem: 

- Adoptium template updated to cover EA and GA cases:
  - Support of JDK8, 11, 17, 21 and 21-preview
  - Support of EA and GA
  - Support of "1 digit only" versions such as `21+35`
  - Better management of line returns (allowing to remove a now-unneeded `chop()` for `build::agent` profile)
  - A LOT of comments added to explain
  - Vagrant hieradata setup to cover a lot of cases

- Fixed a bug for `build::agent` profile where `arm64` / `aarch64` architectures were not supported for JDK installation
- Updated the `check-jdk.sh` shell script used by updatecli, cherry-picking from jenkinsci/docker where the split "21 GA/ 21 EA" already exists

- As only the ci.jenkins.io's `s390x` agent need a JDK21 EA, the EA version is specified only for it, otherwise JDK21 GA is used everywhere else
  - Added an updatecli manifest to track only the EA version for `s390x` agent

The following tests were run locally, in addition of to the puppet unit tests:

- `vagrant up --provision jenkins::agent` => check it installs the proper JDK versions (tried on a macOS machine with Silicon so the `aarch64` temurin binaries were installed, and tried forcing the `amd64` which downloaded the x86 binaries as expected)
- `vagrant up --provision jenkins::controller` the 2 new cases (EA 3 digits and EA 1 digit):
  - Provisionning went well
  - The following command returned `200` for each of the found links: `for url in $(cat /var/lib/jenkins/casd.d/tools.yaml | grep temurin | cut -d'"' -f2);do echo "== echo $url"; curl -fsSL -I -o /dev/null -w "%{http_code}" "$url"; echo; done`

- Ran the new updatecli manifest (for 21 EA) locally with the target's `scmdid` commented out: it returned `⚠ - key "$.profile::jenkinscontroller::jcasc.tools.jdk.jdk21.installers.s390x.version" should be updated from "21+35-ea-beta" to "21.0.1+12-ea-beta", in file "hieradata/clients/controller.ci.jenkins.io.yaml"`

File: updatecli/scripts/check-jdk.sh
Comment (dduportal) [reply updatecli/scripts/check-jdk.sh:58]: For info, see https://github.com/jenkins-infra/jenkins-infra/pull/3113#issuecomment-1786727093

**Comment by dduportal** [comment]
Ready for review (cc @hervelemeur):

- Please don't merge it, as I would want to deploy it using `noop` and carefull testing on *.ci.jenkins.io
- The updatecli check failed due to the new `s390x` key not found: will be successfuly once merged (see my comment above: was tested)
- The JDK21 version is still pinned to 21+35(.*). Updatecli is expected, once merged, to bump to 21.0.12(.*).

File: updatecli/scripts/check-jdk.sh
Comment (smerle33) [inline updatecli/scripts/check-jdk.sh]: jdk 19 ?

File: updatecli/scripts/check-jdk.sh
Comment (smerle33) [inline updatecli/scripts/check-jdk.sh]: jdk19 ?

**Review by smerle33 — COMMENTED** [review]
except for the JDK19, it looks perfect.

File: updatecli/scripts/check-jdk.sh
Comment (dduportal) [reply updatecli/scripts/check-jdk.sh]: Good catch! Removed in https://github.com/jenkins-infra/jenkins-infra/pull/3113/commits/29cd4d756c40b42af68929c9c558bbe45ecf6193

File: .gitignore
Comment (dduportal) [inline .gitignore:14]: Added because local tests are creating a "rclone.zip" file at the root of the repository. Will be fixed in a later PR

File: updatecli/scripts/check-jdk.sh
Comment (dduportal) [reply updatecli/scripts/check-jdk.sh]: Good catch! Removed in https://github.com/jenkins-infra/jenkins-infra/pull/3113/commits/29cd4d756c40b42af68929c9c558bbe45ecf6193

**Review by lemeurherve — APPROVED** [review]
Nice 👏

**Comment by dduportal** [comment]
Merging to deploy and verify in its current state

**Comment by dduportal** [comment]
- After deployment, run the following  pipeline with success:

```
parallel linuxs390x: {
    node('s390x') {
        infra.runMaven(['-v'], 21)
    }
},
linuxamd64vm: {
    node('linux-amd64 && vm') {
        infra.runMaven(['-v'], 21)
    }
},
linuxamd64container: {
    node('kubernetes') {
        infra.runMaven(['-v'], 21)
    }
},
windowsamd64vm: {
    node('windows && vm') {
        infra.runMaven(['-v'], 21)
    }
},
windowsamd64container: {
    node('aci') {
        infra.runMaven(['-v'], 21)
    }
}
```

- 2 PRs were opened (https://github.com/jenkins-infra/jenkins-infra/pull/3165/files and https://github.com/jenkins-infra/jenkins-infra/pull/3166/files) as expected: proof this PR works as expected

**Pull request merged**
